### PR TITLE
Initial support for `LIKE` in Cypher.

### PR DESCRIPTION
**Pull request by pontusmelke**
Implemented `LIKE` as a syntactic sugar for using regular expression.
Support `%`, `_` and `[]` in the `LIKE` expression.

Usage:
```
MATCH (a) WHERE a.name LIKE 'ABCDEF' RETURN a
MATCH (a) WHERE a.name NOT LIKE 'ABCDEF' RETURN a
MATCH (a) WHERE a.name ILIKE 'ABCDEF' RETURN a
```

### CODE REVIEW

**Comment by jexp** [comment]
Is ILIKE part of this, i.e. case insensitive matching? Should be easy to add with `(?:i)`
Definitely something that is requested often in real world data.

**Comment by craigtaverner** [comment]
Took a peek at the code and tests and it looks easy to add, but curiously there are tests that are case sensitive, so we'd better re-think those.

**Comment by pontusmelke** [comment]
@jexp no this contains only LIKE but ILIKE shouldn't be too hard to add
@craigtaverner LIKE is case sensitive so the tests should also be that, ILIKE is a separate operator

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+//TODO should option be supported? In standard?
+/**
+ * Parser that parses a Like expression into [[ParsedLike]] which contains tokens of [[LikeOp]].
+ *
+ * A Like expression can contain the special characters:
+ *  - % wildcard
+ *  - _ matches exactly one character
+ *  - [XY] optionally match any of the characters
+ */
+case object LikeParser extends Parser {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala]: How do we make sure to report the correct line, column, offsets in errors by using an inner parser?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/Predicate.scala [hunk 2]
```diff
@@ -251,8 +251,8 @@ case class PropertyExists(identifier: Expression, propertyKey: KeyToken) extends
   override def localEffects = Effects.READS_ENTITIES
 }
 
-case class LiteralRegularExpression(a: Expression, regex: Literal) extends Predicate {
-  lazy val pattern = regex.v.asInstanceOf[String].r.pattern
+case class LiteralRegularExpression(a: Expression, regex: Literal)(implicit converter: String => String = ((a) => a)) extends Predicate {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/Predicate.scala]: `regex: Literal` -> `regex: StringLiteral`? Or perhaps even `regex: RegexLiteral`?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala [hunk 3]
```diff
@@ -47,6 +47,8 @@ object ExpressionConverters {
       case e: ast.NotEquals => e.asCommandNotEquals
       case e: ast.RegexMatch => e.asCommandRegex
       case e: ast.In => e.asCommandIn
+      case e: ast.Like => e.asCommandRegex
+      case e: ast.NotLike => e.asCommandNegatedRegex
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala]: I think we should not compile `NotLike` to commands but rewrite it to `Not(Like)` as for `NotEquals` -> `Not(Equals)`.

The compilation of `NotEquals` to commands is still there because of legacy compatibility, Ronja does not uses it.

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convertLikeToRegex.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_2.ast
+
+import java.util.regex.Pattern.quote
+
+/**
+ * Converts [[ParsedLike]] into a regular expression string
+ */
+case object convertLikeToRegex {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convertLikeToRegex.scala]: Why is this object in `ast` package?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_2.ast
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala]: Should this leave in `parser` rather than `ast`?

**Comment by thobe** [comment]
I know SQL has 'LIKE', but I'm not convinced it's a great idea to have multiple ways in the language to do string matching. We already have regex match. Would it really be that hard to parse the regex patterns and find out the opportunities for optimisation we can gain from them by turning it into prefix lookups et.c.

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala (hunk 1 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala]: I would assume with a full regex parser this would be a problem, but with only '%' and '_' wildcards, there does not seem to be any way the internal parsing can generate an error (other than a bug in our code, of course). So is it not sufficient to return the offset of the beginning of the pattern string on any inner parser errors?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala (hunk 3 above)
Comment (craigtaverner) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala]: Since "a NOT LIKE b" is infix, we supported it as its own expression. Perhaps a rewriter could be considered. Certainly the existing code has "a NOT LIKE b" behaving the same as "NOT(a LIKE b)" with a different parsing chain. They do generate the same AST though. Perhaps that is good enough? See CypherParserTest.scala#316

**Comment by craigtaverner** [comment]
The current code seems to deviate from the standard in two ways:

* The support for the [abc] set match is not mentioned in the standard (I think we should drop it)
* The standard describes "a LIKE b ESCAPE x", which we do not support

I think we should drop support for the character set match and add support for ESCAPE. The good thing with that is that the character set match is a more complex regex which could (I imagine) be a source of possible parsing errors (highlighting @davidegrohmann concern), while the ESCAPE character can be handled easily at a high level in the regex writing (and in future also if we re-write to lucene queries).

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala (hunk 1 above)
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/LikeParser.scala]: Sure, we have few operator and the syntax is simple, but this is the current situation. There is no guarantee it will be so simple forever, moreover having consistent error messages is quite nice. 

On the other hand, we have already a quite monolithic parser, i.e., patterns, expressions, clauses are all parsed by the same parser, so I have a bit of hard time to see the benefit of having a separate parser for this use case.

**Comment by craigtaverner** [comment]
@thobe, pros of LIKE:

* Everyone knows how it works (least surprise)
* Optimizing for lucene or other backend should be easy due to very limited range of possible patterns

Of course it would also be very nice (but perhaps no that easy) to also optimize regex onto lucene (or other backend).

**Comment by jexp** [comment]
we can also optimize for things like `s.startsWith(), endsWith() and contains()` so we don't need regexps at all.

I think there is [experimental regexp support](https://lucene.apache.org/core/4_0_0/core/org/apache/lucene/util/automaton/RegExp.html?is-external=true) for lucene in 4.x.

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala (hunk 3 above)
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala]: I am not against having `LIKE` and `NOT LIKE` at the AST level, but before planning we should remove the syntactic sugar as we do for `Equals`. I am not sure I got the same-generate-expression part of your comment since I have in mind the classes `ast.Like` and `ast.NotLike` as by diff above.

**Comment by jexp** [comment]
I've never seen or used ESCAPE in my whole life. Not sure how relevant it is.

**Comment by thobe** [comment]
@craigtaverner I think you mean "~~Everyone~~*Many people* **who know SQL** know how it works".

We already have a problem with regex in that we don't own the regex syntax, we just pass the string on to some third party regex engine. I think we really should compile regular expressions on our own as well, in order to own the syntax, even if we compile to a third party regex engine. And if we invest that effort into parsing regular expressions the additional work to support using and optimising them throughout the query execution stack would be minimal.

The other approach (which might be equally nice), that @davidegrohmann pointed out to me, is to drop support for regex matching with `~=` and only support `LIKE`.

**Comment by danielcor** [comment]
Another point -- different vendors support LIKE in SQL differently, and
people's awareness of the language depends on which database they have used.

Microsoft supports: %, _, [], and [^] as well as ESCAPE.
  It requires a TOUPPER(x) to be case insensitive
  (details: https://msdn.microsoft.com/en-us/library/ms179859.aspx)

MySQL supports %, _  and not [], [^] OR ESCAPE.   It is always case
insensitive.
  (details: http://dev.mysql.com/doc/refman/5.0/en/pattern-matching.html)

Oracle implementation issimilar to MySQL - except by default it's case
sensitive like Microsoft:
http://docs.oracle.com/cd/B12037_01/server.101/b10759/conditions016.htm

It also supports LIKEC, LIKE2, LIKE4 - to handle unicode character sets in
different ways.



On Thu, Jan 22, 2015 at 11:10 AM, Tobias Lindaaker <notifications@github.com

**Comment by pontusmelke** [comment]
@davidegrohmann The benefit of having a separate parser is that we need to be able parse the 'LIKE expression' at runtime (e.g. when provided as parameter) as well as when parsing the actual CYPHER query. Using the entire cypher parser machinery at runtime seems like a bad idea. I still need to look at the support positions for error messages but I guess we might end up in situation where we only can give the position of the 'LIKE expression' in the error message and not the position within the actual string.

@danielcor That's really nice input, I was just trying to figure out who supports []…

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala (hunk 3 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/convert/commands/ExpressionConverters.scala]: @davidegrohmann The `NotEquals` -> `Not(Equals)` transformation is done on the commands right? There is no `Like` on the command level (yet?). We transform to `RegularExpression` and `Not(RegularExpression)` so I am not sure exactly what kind of rewrite you propose?

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/Predicate.scala (hunk 2 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/Predicate.scala]: @davidegrohmann I didn't change anything here apart from the converter function? However I am not crazy about the cast and I am not sure why `pattern` is lazy so refactoring might be a good idea.

**Comment by systay** [comment]
I'll close this PR for now. We will iterate on the language spec to take the feedback into account, and will re-open this PR when that work is ready.
